### PR TITLE
Switch to C++17.

### PR DESCRIPTION
**Pull request by mpreiner**

### CODE REVIEW

**Review by nafur — APPROVED** [review]
LGTM. Yay! 🎉

File: src/expr/node.h [hunk 1]
```diff
@@ -1451,8 +1451,13 @@ NodeTemplate<ref_count>::substitute(Iterator substitutionsBegin,
...
   // otherwise compute
-  Iterator j = find_if(substitutionsBegin, substitutionsEnd,
-                       bind2nd(first_equal_to<typename Iterator::value_type::first_type, typename Iterator::value_type::second_type>(), *this));
+  Iterator j = find_if(
+      substitutionsBegin,
+      substitutionsEnd,
+      std::bind(first_equal_to<typename Iterator::value_type::first_type,
+                               typename Iterator::value_type::second_type>(),
+                std::placeholders::_1,
+                *this));
```
Comment (nafur) [inline src/expr/node.h:1460]: How about this?
```suggestion
  Iterator j = find_if(
      substitutionsBegin,
      substitutionsEnd,
      [this](const auto& subst){ return subst.first == *this; });
```

**Review by nafur — CHANGES_REQUESTED** [review]
My feeling is that lambdas are usually superior to std::bind in modern C++.
Not only are they far more readable, they are also usually more efficient: lambdas end up to be regular functions that can be inlined.

**Comment by nafur** [comment]
So the next issue: dynamic exception specification was removed with c++17 (i.e. `void f() throw(...)`). As far as I can tell, we didn't use them anyway, except for a few commented out instances in `delta_rational.h`. With this removal, `std::set_unexpected` et al were removed as well, and we can simply remove them from our codebase. Also, we might want to consider marking functions `noexcept`.

**Review by nafur — APPROVED** [review]
LGTM.

**Comment by ajreynol** [comment]
We should performance test SMT-LIB quantifiers, where substitute can be a bottleneck

**Comment by ajreynol** [comment]
Nevermind, the changed variant of substitute is not the commonly used one.

**Pull request merged**